syntax = "proto3";

package light; 


// Service for controlling lights
service LightService {
    // unary
    rpc GetLight (Indicator) returns (LightResponse) {}
    rpc SetLight (LightRequest) returns (Empty) {}

    // server --> client streaming
    rpc GetLights (Empty) returns (stream LightResponse) {}
    
    // client --> server streaming
    rpc SetLights (stream LightRequest) returns (Empty) {}

    // client <--> server streaming
    rpc CheckLights (stream Indicator) returns (stream ChatMessage) {}
}


// Message for a light object
message Light {
    uint32 id = 1;
    bool on = 2;
    uint32 red = 3;
    uint32 green = 4;
    uint32 blue = 5;
}

// Message for a light request
message LightRequest {
    Light light = 1;
}

// Message for a light response
message LightResponse {
    string message = 1;
    Light light = 2;
}

// Message for an indicator
message Indicator {
    uint32 id = 1;
}

// Empty message
message Empty {}

// Chat message 
message ChatMessage {
    uint32 id = 1;
    string message = 2;
}


// run with "python3 -m grpc_tools.protoc -I./protos protos/light.proto --python_out=. --grpc_python_out=."